### PR TITLE
feat(common): Remove uuid and update unique id generation

### PR DESCRIPTION
**Pull request by NicholasBoll**
## Summary

Our current id generation has issues.

- It generates very long ids that add noise to the page
- It is possible to generate a blank id
- Id generation is slow
- Snapshot tests require mocking the `uuid` module

Fixes: #1311, #1106, #664, #602

category

### Release Note
**NOTE for jest snapshots**: This change removes the `uuid` package and instead will generate a one-time client seed and then create auto-incrementing ids. This change will not break UI or automated UI tests. It will break snapshot tests however. Previously, the only way to get stable ids for snapshot tests was to mock the `uuid` module. This was an implementation detail. To make snapshots work again, add the following to your jest setup file:

```ts
import {setUniqueSeed, resetUniqueIdCount} from '@workday/canvas-kit-react/common';

beforeEach(() => {
  setUniqueSeed('a'); // force set the seed
  resetUniqueIdCount(); // reset the unique id count
});
```

This will ensure each Jest snapshot has ids that look like `a0` and `a1` and will be the same every time the snapshot is run. Do not use these methods in production though - it may lead to inaccessible applications due to IDREF collisions.

---

## Additional References

Test of different ID generation methods https://codesandbox.io/s/tkod6

Remaining dependencies on older versions of uuid:
- [x] `request` - this is a deprecated lib that we use for scripts
- [ ] `node-sass` - We're using v4. Newest is v7. This PR removed `request` which is using `uuid@3`, but hasn't been released yet: https://github.com/sass/node-sass/pull/3193. Currently blocked. `sass-loader` that is compatible with `node-sass@7` requires `webpack@5` which Storybook doesn't currently support.
- [x] `jest-unit` - We're using v10. Current version is v13
- [ ] `lerna` - We're using v3. Current version is v4
- [ ] `react-docgen-typescript-loader` - We're on v3.3.0. Latest is v3.7.2
- [ ] `storybook/addon-docs` - Oh, Storybook...

### CODE REVIEW

File: cypress/integration/Combobox.spec.ts [hunk 1]
```diff
@@ -37,7 +37,7 @@ describe('Combobox', () => {
     });
 
     it('should not show menu', () => {
-      cy.findByRole('listbox').should('not.be.visible');
+      cy.findByRole('listbox').should('not.exist');
```
Comment (NicholasBoll) [inline cypress/integration/Combobox.spec.ts:40]: All these were due to a breaking change in [Cypress v6](https://docs.cypress.io/guides/references/changelog#6-0-0).

This "fix" did find legitimate issues with our tests, but I feel like we've lost some semantics. I want to know if the element is no longer accessible to the user. Whether it is still in the DOM or just hidden is an implementation detail. Perhaps I'll make an `be.inaccessible` assertion that does this.

File: cypress/integration/Tooltip.spec.ts [hunk 2]
```diff
@@ -155,14 +155,14 @@ describe('Tooltip', () => {
     });
 
     it('the span element should not have an aria-describedby attribute', () => {
-      cy.get('button').should('not.have.attr', 'aria-describedby');
+      cy.get('span').should('not.have.attr', 'aria-describedby');
```
Comment (NicholasBoll) [inline cypress/integration/Tooltip.spec.ts:158]: This was caught by the change in Cypress v6. Before, `not.have.attr` would pass even if `cy.get('button')` never actually matched anything. This was a typo from long ago that Cypress was perfectly happy to incorrectly pass.

File: cypress/integration/Tooltip.spec.ts [hunk 3]
```diff
@@ -155,14 +155,14 @@ describe('Tooltip', () => {
...
     });
 
     it('the span element should not have an aria-label attribute', () => {
-      cy.get('button').should('not.have.attr', 'aria-describedby');
+      cy.get('span').should('not.have.attr', 'aria-describedby');
     });
 
-    context('when the "Delete" button is hovered', () => {
+    context('when the "Some Text" text is hovered', () => {
```
Comment (NicholasBoll) [inline cypress/integration/Tooltip.spec.ts:165]: This was a copy/paste error. The tooltip isn't from a delete button, but a span with some text.

File: cypress/support/commands.ts [hunk 4]
```diff
@@ -19,24 +19,33 @@ Cypress.Commands.add('injectAxe', () => {
 });
 
 // Needed for https://github.com/Bkucera/cypress-plugin-tab/issues/46
-Cypress.Commands.overwrite('visit', (originalFn, url, options = {}) => {
-  if (typeof url === 'object') {
-    // eslint-disable-next-line no-param-reassign
-    url = options.url;
+Cypress.Commands.overwrite('visit', (originalFn, urlOrOptions, inputOptions = {}) => {
```
Comment (NicholasBoll) [inline cypress/support/commands.ts:22]: Changes in this file were from Cypress adding type defintions to `Cypress.Commands.add` and `Cypress.Commands.overwrite` in [Cypress v9](https://docs.cypress.io/guides/references/changelog#9-0-0). In most cases, this is an improvement. In the case of commands that use Typescript overrides, this causes problems. I've rearranged command overrides and added some `any` in cases where the Cypress types introduce problems that cannot be resolved.

File: cypress/support/cypress-plugin-tab.d.ts [hunk 5]
```diff
@@ -1,5 +1,5 @@
 declare namespace Cypress {
-  interface Chainable {
-    tab(options?: Partial<{shift: Boolean}>): Chainable;
+  interface Chainable<Subject = any> {
```
Comment (NicholasBoll) [inline cypress/support/cypress-plugin-tab.d.ts:2]: This was the correct typing all along

File: cypress/support/index.ts [hunk 6]
```diff
@@ -6,3 +6,11 @@ import 'cypress-keyboard-plugin';
...
 
 import './commands';
+
+const resizeObserverLoopErrRe = /^[^(ResizeObserver loop limit exceeded)]/;
+Cypress.on('uncaught:exception', err => {
+  /* returning false here prevents Cypress from failing the test */
+  if (resizeObserverLoopErrRe.test(err.message)) {
+    return false;
+  }
+});
```
Comment (NicholasBoll) [inline cypress/support/index.ts:16]: The `ResizeObserver loop limit exceeded` error started happening after upgrading Cypress from `5.6.0` to `9.2.0`. I'm not sure which Cypress version introduced the error. Perhaps later we can find out.

File: jest/setupTests.ts [hunk 7]
```diff
@@ -14,3 +16,8 @@ configure({adapter: new Adapter()});
 // add convenience variables to the global context
 (global as any).verifyComponent = verifyComponent;
 (global as any).jest = jest;
+
+beforeEach(() => {
+  setUniqueSeed('a');
+  resetUniqueIdCount();
+});
```
Comment (NicholasBoll) [inline jest/setupTests.ts:23]: This isn't actually needed for our tests, but it is nice to have something to point to as an example for teams that need this setup.

File: modules/preview-react/index.ts [hunk 8]
```diff
@@ -6,3 +6,5 @@ export * from './select';
 export * from './side-panel';
 export * from './text-area';
 export * from './text-input';
+
+/// <reference types="@types/node" />
```
Comment (NicholasBoll) [inline modules/preview-react/index.ts:10]: This removes Typescript errors complaining that `process` doesn't exist. `uuid` added `@types/node` to the global namespace for us.

File: jest/setupTests.ts (hunk 7 above)
Comment (mannycarrera4) [reply jest/setupTests.ts:24]: I'm sure teams are gonna wanna use this, nice addition

Comment (mannycarrera4) [reply jest/setupTests.ts:24]: Should something like this go into the testing docs for reference?

**Review by mannycarrera4 — COMMENTED** [review]
Looks good to me just some minor comments/ questions

File: modules/react/common/lib/utils/useUniqueId.ts [hunk 9]
```diff
@@ -1,8 +1,24 @@
-import uuid from 'uuid/v4';
-
 import {useConstant} from './useConstant';
 
-export const generateUniqueId = () => uuid().replace(/^[0-9\-]*/gi, '');
+// Create a unique seed per import to prevent collisions from other versions of `useUniqueId`
+let seed = Math.random()
+  .toString(36)
+  .slice(2)
+  .replace(/[0-9]*/, '')
```
[the remark this answers is not in the document]
Comment (NicholasBoll) [reply modules/react/common/lib/utils/useUniqueId.ts:7]: I added it to the `uuid` generation. Base 36 doesn't include dashes.

File: jest/setupTests.ts (hunk 7 above)
Comment (mannycarrera4) [reply jest/setupTests.ts:24]: nvm you documented this pretty well in the readme

**Review by mannycarrera4 — APPROVED** [review]

**Pull request merged**